### PR TITLE
8265066: Split ReservedSpace constructor to avoid default parameter

### PR DESCRIPTION
**Pull request by kstefanj**
Please review this change to change this constructor in ReservedSpace:
```
  // Initialize the reserved space with the given size. If preferred_page_size
  // is set, use this as minimum page size/alignment. This may waste some space
  // if the given size is not aligned to that value, as the reservation will be
  // aligned up to the final alignment in this case.
  ReservedSpace(size_t size, size_t preferred_page_size = 0);
```

I propose to split this constructor into two instead of having the default parameter. This makes the implementation more straight forward. I also make the single argument constructor explicit to avoid implicit conversion. There was one place where we relied on implicit conversion and this has been updated to explicitly create the ReservedSpace.

This cleanup is another step towards:
[JDK-8261527: Record page size used for underlying mapping in ReservedSpace](https://bugs.openjdk.java.net/browse/JDK-8261527)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265066](https://bugs.openjdk.java.net/browse/JDK-8265066): Split ReservedSpace constructor to avoid default parameter

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3457/head:pull/3457` \
`$ git checkout pull/3457`

Update a local copy of the PR: \
`$ git checkout pull/3457` \
`$ git pull https://git.openjdk.java.net/jdk pull/3457/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3457`

View PR using the GUI difftool: \
`$ git pr show -t 3457`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3457.diff">https://git.openjdk.java.net/jdk/pull/3457.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/memory/virtualspace.cpp [hunk 1]
```diff
@@ -44,21 +44,27 @@ ReservedSpace::ReservedSpace() : _base(NULL), _size(0), _noaccess_prefix(0),
...
-  size_t alignment;
-  if (large_pages && has_preferred_page_size) {
-    alignment = MAX2(page_size, (size_t)os::vm_allocation_granularity());
-    // ReservedSpace initialization requires size to be aligned to the given
-    // alignment. Align the size up.
+  initialize(size, alignment, large_pages, NULL, false);
+}
+
+ReservedSpace::ReservedSpace(size_t size, size_t preferred_page_size) : _fd_for_heap(-1) {
+  assert(is_power_of_2(preferred_page_size), "invariant");
```
[the remark this answers is not in the document]
Comment (kstefanj) [reply src/hotspot/share/memory/virtualspace.cpp:58]: Yes, you are correct. The assert would only catch bogus values, and having the assert query the os-layer here might be a bit over the top. I'll just remove the assert for now.

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Review by stefank — APPROVED** [review]

**Comment by kstefanj** [comment]
Thanks for the reviews @tschatzl and @stefank.

/integrate

**Pull request closed without merging**